gnosis-imap/gnosis-imap-spec.ini

```diff
@@ -72,7 +72,137 @@ localMem: M
+    64,                     32, _),
+
+    ;start of result, divided into 4 parts:
+    ;return abi.encodePacked(byte(0x19), byte(0x01), domainSeparator, safeTxHash)
+    INIT_USED_MEM_PTR +Int 416, 32, #padRightToWidth(32, #parseByteStack("0x19")) ),
+    INIT_USED_MEM_PTR +Int 417, 32, #padRightToWidth(32, #parseByteStack("0x01")) ),
+    INIT_USED_MEM_PTR +Int 418, 32, #buf(32, DOMAIN_SEPARATOR) ),
+    INIT_USED_MEM_PTR +Int 450, 32, #buf(32, SAFE_TX_HASH ) ),
+    ;output len
+    INIT_USED_MEM_PTR +Int 384, 32, #padToWidth(32, #asByteStack(66)) ),
+    ;next free memory loc
+    64,                     32, #buf(32, INIT_USED_MEM_PTR +Int 482) )
+pc: 16193 => 16786
+gas: #gas(STARTGAS,
+          X                                                => (X +Int 6 *Int ((DATA_LEN +Int 31) /Int 32)) +Int 941,
+          Cmem(BYZANTIUM, START_MEM_USED => END_MEM_USED ) )
+memoryUsed: START_MEM_USED => END_MEM_USED
+log: _
+refund: _
+coinbase: _
+proxy_storage: STORAGE
++requires:
+    ; Range
+    andBool #rangeAddress(MSG_SENDER)
+
+    andBool #rangeAddress(          TO)
+    andBool #rangeUInt( 256,        VALUE)
+    andBool #rangeUInt( 256,        DATA_LEN)
+    ;andBool #rangeBytes( DATA_LEN,  DATA)
+    andBool #rangeUInt(   8,        OPERATION)
+    ; enum Enum.Operation, 3 possible values encoded to 0-2.
+    andBool OPERATION <=Int 2
+    andBool #rangeUInt( 256,        SAFE_TX_GAS)
+    andBool #rangeUInt( 256,        DATA_GAS)
+    andBool #rangeUInt( 256,        GAS_PRICE)
+    andBool #rangeAddress(          GAS_TOKEN)
+    andBool #rangeAddress(          REFUND_RECEIVER)
+    andBool #rangeUInt( 256,        NONCE)
+
+    andBool #rangeBytes( 32,        DOMAIN_SEPARATOR)
+
+    ; TODO more precise value. 1000 is too little.
+    andBool #range(0 <= #sizeWordStack(WS) <= 900)
+
+    andBool DATA_LEN <Int 2 ^Int 16
+
+    andBool SAFE_TX_HASH ==Int keccak( #encodeArgs(
+                                 #bytes32(#parseHexWord({SAFE_TX_TYPEHASH})),
+                                 #address(TO),
+                                 #uint256(VALUE),
+                                 #bytes32(keccak(#buf(DATA_LEN, DATA))),
+                                 #uint8(OPERATION),
+                                 #uint256(SAFE_TX_GAS),
+                                 #uint256(DATA_GAS),
+                                 #uint256(GAS_PRICE),
+                                 #address(GAS_TOKEN),
+                                 #address(REFUND_RECEIVER),
+                                 #uint256(NONCE) ))
+    andBool INIT_MEM ==K   storeRange( storeRange( storeRange( storeRange( _,
+                              64,                   32,       #buf(32, INIT_USED_MEM_PTR) ),
+                              DATA_LEN_LOC,         32,       #buf(32, DATA_LEN) ),
+                              DATA_BUF_LOC,         DATA_LEN, #buf(DATA_LEN, DATA) ),
+                              INIT_USED_MEM_ACTUAL, 32,       #padToWidth(32, #asByteStack(0)) )
```

Comment:
in the trusted spec, we should have the following instead of the above:
```
INIT_USED_MEM_PTR  ==Int #asWord(selectRange(INIT_MEM, 64, 32))
DATA_LEN ==Int #asWord(selectRange(INIT_MEM, DATA_LEN_LOC, 32))
DATA_BUF ==K selectRange(INIT_MEM, DATA_BUF_LOC, DATA_LEN)
```
where we replace `#buf(DATA_LEN, DATA)` by `DATA_BUF` in the above `SAFE_TX_HASH` equality.